optimum/intel/openvino/modeling_diffusion.py

```diff
@@ -829,24 +987,26 @@ def forward(
         return_dict: bool = False,
     ):
         self._compile()
-
         model_inputs = {"input_ids": input_ids}
 
-        ov_outputs = self.request(model_inputs, share_inputs=True).to_dict()
-
+        ov_outputs = self.request(model_inputs, share_inputs=True)
+        main_out = ov_outputs[0]
         model_outputs = {}
-        for key, value in ov_outputs.items():
-            model_outputs[next(iter(key.names))] = torch.from_numpy(value)
-
-        if output_hidden_states:
-            model_outputs["hidden_states"] = []
-            for i in range(self.config.num_hidden_layers):
-                model_outputs["hidden_states"].append(model_outputs.pop(f"hidden_states.{i}"))
-            model_outputs["hidden_states"].append(model_outputs.get("last_hidden_state"))
+        model_outputs[self.model.outputs[0].get_any_name()] = torch.from_numpy(main_out)
+        if len(self.model.outputs) > 1 and "pooler_output" in self.model.outputs[1].get_any_name():
+            model_outputs["pooler_output"] = torch.from_numpy(ov_outputs[1])
+        if self.hidden_states_output_names and "last_hidden_state" not in model_outputs:
+            model_outputs["last_hidden_state"] = torch.from_numpy(ov_outputs[self.hidden_states_output_names[-1]])
+        if (
+            self.hidden_states_output_names
+            and output_hidden_states
+            or getattr(self.config, "output_hidden_states", False)
+        ):
+            hidden_states = [torch.from_numpy(ov_outputs[out_name]) for out_name in self.hidden_states_output_names]
+            model_outputs["hidden_states"] = hidden_states
```

Comment:
not sure this complexity is needed, the original code does the same thing by seperating the `torch.from_numpy` and collecting the hidden states in a list if requested, the only change you'll have to do is to accounf for the missing `self.config.num_hidden_layers` by using `self.config.num_layers` otherwise for t5-encoder https://huggingface.co/stabilityai/stable-diffusion-3-medium-diffusers/blob/main/text_encoder_3/config.json#L21

Reply:
problem that in some cases torch.from_numpy doing coping of tensors that is not good from memory usage point of view. Additonally, in torch hidden states outputs provided as list, while openvino has then flattenized, so it may lead to additional issues and unexpected keys in ModelOutput (some pipelines uses indexing for accessing outputs and indexation will not the same like in original output format). Also, there is possible issue that output with name last_hidden_state will not in outputs (due to renaming output ports binded to the same tensor it exists in output twice as last_hidden_stat and the last output from hidden_states list, in ov such tensors will have the same name and in some cases it will be hidden_states.n instead of last_hidden_state

Reply:
so I tried to make output closer to original one for avoid issues in future, but if you do not agree, please say something. that is not problem for me to use your suggested way

Reply:
> please say something

I already did. but I have no issue with it if you're satisfied with it, it does the job.

Reply:
> I already did.

you did that before I shared my opinion, in my perception, code review is dialog and if it requires some discussion I need to receive feedback on my answer too, because I need to know what do you think about my point of view and I need to change something after my explanation or not. That will be very helpful as it is normal that some decision may changed during discussion.

Reply:
it's too much wasted time

> problem that in some cases torch.from_numpy doing coping of tensors that is not good from memory usage point of view if we will not care about necessarily or unnecessarily of wrapping specific output.

not true, check documentation https://pytorch.org/docs/stable/generated/torch.from_numpy.html, and this suggestion does the same, just in different places.

> Additonally, in torch hidden states outputs provided as grouped list, while openvino has them flattened, so it may lead to additional issues and unexpected keys in ModelOutput (some pipelines uses indexing for accessing outputs and indexation will not the same like in original output format).

I believe that's why we have model configs with inputs/outputs specifications, we also control the order.

> Also, there is possible issue that output with name last_hidden_state will not in outputs (due to renaming output ports binded to the same tensor, when it exists in output twice as last_hidden_stat and the last output from hidden_states list, in ov such tensors will have the same name and in some cases it will be hidden_states.n instead of last_hidden_state. as usage last_hidden_state is important for some pipelines, I'm trying to restore that in original format

I don't get this part, I really think last_hidden_state will be there since in order, if it's there it comes before the hidden_states list, it's the hidden_states.n that might be missing in this case but Idk this sounds like an issue that requires a minimal reproducible code, otherwise I haven't seen it before.

Reply:
I do not want to spend your time in long additional explanation how openvino works, all things that you said does not take into account product realization and its specific, so please trust me as an engineer that I also have some experiments and know more about potential pitfalls, not just hypothetical and theoretical assumptions